src/legacy/core_plugins/embeddable_api/public/containers/container.ts

```diff
@@ -0,0 +1,381 @@
+    // by the listener.
+    const panels = { ...this.input.panels };
+    delete panels[embeddableId];
+    this.updateInput({ panels } as Partial<TContainerInput>);
+  }
+
+  public getChild<E extends IEmbeddable>(id: string): E {
+    return this.children[id] as E;
+  }
+
+  public getInputForChild<TEmbeddableInput extends EmbeddableInput = EmbeddableInput>(
+    embeddableId: string
+  ): TEmbeddableInput {
+    const containerInput: TChildInput = this.getInheritedInput(embeddableId);
+    const panelState = this.getPanelState(embeddableId);
+
+    const explicitInput = panelState.explicitInput;
+    const explicitFiltered: { [key: string]: unknown } = {};
+
+    const keys = getKeys(panelState.explicitInput);
+
+    // If explicit input for a particular value is undefined, and container has that input defined,
+    // we will use the inherited container input. This way children can set a value to undefined in order
+    // to default back to inherited input. However, if the particular value is not part of the container, then
+    // the caller may be trying to explicitly tell the child to clear out a given value, so in that case, we want
+    // to pass it along.
+    keys.forEach(key => {
+      if (explicitInput[key] === undefined && containerInput[key] !== undefined) {
+        return;
+      }
+      explicitFiltered[key] = explicitInput[key];
+    });
+
+    return ({
+      ...containerInput,
+      ...explicitFiltered,
+      // Typescript has difficulties with inferring this type but it is accurate with all
+      // tests I tried. Could probably be revisted with future releases of TS to see if
+      // it can accurately infer the type.
+    } as unknown) as TEmbeddableInput;
+  }
+
+  public destroy() {
+    super.destroy();
+    Object.values(this.children).forEach(child => child.destroy());
+    this.subscription.unsubscribe();
+  }
+
+  public async untilEmbeddableLoaded<TEmbeddable extends IEmbeddable>(
+    id: string
+  ): Promise<TEmbeddable | ErrorEmbeddable> {
+    return new Promise(resolve => {
+      const subscription = this.getOutput$().subscribe(() => {
+        if (this.output.embeddableLoaded[id]) {
+          subscription.unsubscribe();
+          resolve(this.children[id] as TEmbeddable);
+        }
+      });
+    });
+  }
+
+  protected createNewPanelState<
+    TEmbeddableInput extends EmbeddableInput,
+    TEmbeddable extends IEmbeddable<TEmbeddableInput, any>
+  >(
+    factory: EmbeddableFactory<TEmbeddableInput, any, TEmbeddable>,
+    partial: Partial<TEmbeddableInput> = {}
+  ): PanelState<TEmbeddableInput> {
+    const embeddableId = partial.id || uuid.v4();
+
+    const explicitInput = this.createNewExplicitEmbeddableInput<TEmbeddableInput>(
+      embeddableId,
+      factory,
+      partial
+    );
+
+    return {
+      type: factory.type,
+      embeddableId,
+      explicitInput: {
+        id: embeddableId,
```

Comment:
I think this is unneccessary if I make id required in explicit input, I can just always store it there.

Id handling is a little awkward.  Like @streamich  and @lizozom  pointed out earlier, why use a unique id and not just an index into an array of panels because right now all it needs to be is unique within a contianer?  And at that rate, is there any reason to actually require an embeddable to have an id if it's not inside a container?  Not really.  But then it gets a little awkward being like, okay id required if in a container, otherwise not required.  

So this could probably use some improvement but I want to see where phase 2 goes with it... that was also part of what initiated an id... when I tied actionId to embeddable Id in a saved object then realized having non unique embeddable ids would cause problems.  Since phase 2 veered away from actions as saved objects that might not be an issue anymore.  But I want to wait to see what happens with phase 2 before making any big changes with id.

But I think I can get rid of `embeddableId` at the top level to at least avoid the duplication for this PR, and change all references to `panel.explicitInput.id`.